### PR TITLE
User form

### PR DESCRIPTION
**Pull request by hate-bug**
1. Users can submit a form on the web page. 
image
2. Admin can review applicants in the admin page. 
image
3. An email will send to the user when an Approve decision is made. 
4. User can change the password after logging in. 
image
image

### CODE REVIEW

File: resources/static/index.html [hunk 1]
```diff
@@ -16,18 +16,22 @@
             <div id="exit" style="display: none" class="col-4 pt-1 text-left">
                 <a class="text-muted" href="/">Back</a>
             </div>
-            <div class="text-center">
+            <div class="text-cente">
```
Comment (esfandia) [inline resources/static/index.html:19]: "cente" seems to be typo?

File: src/main/java/com/Model/Applicant.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    private long id;
+
+    private String applicantEmail;
+
+    private String description;
+
+    public Applicant(){
```
Comment (esfandia) [inline src/main/java/com/Model/Applicant.java:19]: use this("unknown", "unknown")

File: src/main/java/com/Payload/ConfirmPasswordEntity.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
+package com.Payload;
+
+public class ConfirmPasswordEntity {
```
Comment (esfandia) [inline src/main/java/com/Payload/ConfirmPasswordEntity.java:3]: why is it called an Entity but doesn't have an @Entity? Is this persisted?

File: src/main/java/com/Service/MessageSender.java [hunk 4]
```diff
@@ -18,14 +22,20 @@ private MessageSender (){}
...
+        message.setTo(applicant.getApplicantEmail());
+        message.setSubject("COVID-19 tracking system account.");
         message.setFrom(this.senderEmailAddress);
-        message.setText("To confirm your email address, please click here:"
-                + serverAddress +"/confirmaccount?token="+token);
+        String password = UUID.randomUUID().toString();
+        User user = new User(applicant.getApplicantEmail(), password);
+        this.userRepository.save(user);
+        message.setText("Your account password is:"
+                + password + "\n You can change your password after logged in.");
```
Comment (esfandia) [inline src/main/java/com/Service/MessageSender.java:37]: "after logging in"

File: src/main/java/com/WebController/ApplicationController.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+public class ApplicationController {
+
+    @Autowired
+    private ApplicantRepository applicantRepository;
+
+    @Autowired
+    private MessageSender messageSender;
+
+    @PostMapping (value = "/postapplication")
+    public Applicant postApplication (@RequestBody Applicant applicant) throws FileAlreadyExistsException {
```
Comment (esfandia) [inline src/main/java/com/WebController/ApplicationController.java:23]: is this the right exception to be throwing? We're not dealing with files here.

File: src/main/java/com/WebController/ApplicationController.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        //check if any applicantion already existed
+        if (this.applicantRepository.findAllByApplicantEmailAndDescription(applicant.getApplicantEmail(), applicant.getDescription()).size()==0){
+            return this.applicantRepository.save(applicant);
+        }
+        throw new FileAlreadyExistsException("File exists");
+    }
+
+    @GetMapping (value = "/admin/getapplicants")
```
Comment (esfandia) [inline src/main/java/com/WebController/ApplicationController.java:31]: doesn't the Applicant CrudRepo already provide this endpoint?

File: src/main/java/com/WebController/ApplicationController.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    public Iterable<Applicant> getAllApplicants (@RequestParam("page") int page){
+        return this.applicantRepository.findAll(PageRequest.of(page-1, 10));
+    }
+
+    @PostMapping (value = "/admin/approveapplicant")
```
Comment (esfandia) [inline src/main/java/com/WebController/ApplicationController.java:36]: is this endpoint protected and limited to Admin access?

File: src/test/java/com/Model/ControllerTest/AdminTest.java [hunk 8]
```diff
@@ -52,7 +52,7 @@ public void AdminLoginTest () throws Exception {
         this.userRepository.save(adminUser);
         RequestBuilder requestBuilder = formLogin().user("emailaddress","test@test.test").password("password","testpassword").loginProcessingUrl("/userlogin");
         this.mockMvc.perform(requestBuilder)
-                .andExpect(authenticated().withUsername("test@test.test"));
+                .andExpect(authenticated().withUsername("test@test.test").withRoles("ADMIN", "USER"));
```
Comment (esfandia) [inline src/test/java/com/Model/ControllerTest/AdminTest.java:55]: why is it "ADMIN" in some places and "ADMIN_ROLE" in others?

File: src/main/java/com/Payload/ConfirmPasswordEntity.java (hunk 3 above)
Comment (hate-bug) [reply src/main/java/com/Payload/ConfirmPasswordEntity.java:3]: It's used to Receive the user new password. But not for persistence. I will change the name to ConfirmPasswordPayload.

File: src/main/java/com/WebController/ApplicationController.java (hunk 7 above)
Comment (hate-bug) [reply src/main/java/com/WebController/ApplicationController.java:36]: Yes, in the security configuration class, both "/admin" and "/admin/**" requires ADMIN_ROLE to get access

File: src/main/java/com/WebController/ApplicationController.java (hunk 6 above)
Comment (hate-bug) [reply src/main/java/com/WebController/ApplicationController.java:31]: I will let CrudRepository handle this.

File: src/test/java/com/Model/ControllerTest/AdminTest.java (hunk 8 above)
Comment (hate-bug) [reply src/test/java/com/Model/ControllerTest/AdminTest.java:55]: When testing roles, test methods will automatically and "ROLE_" here.

**Pull request merged**